src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java

```diff
@@ -429,16 +434,20 @@ public String toString() {
 				}
 
 				return sendTask.thenCompose((result) -> {
+					this.responseCommandSent = true;
+					if (this.writeBufferStream != null) {
+						this.writeBufferStream.reset();
+					}
 					this.closed = true;
 					return closeRendezvousAsync();
 				});
 			});
 		}
 
-		CompletableFuture<Void> onWriteBufferFlushTimer() {
-			return this.asyncLock.acquireThenCompose(this.writeTimeout, () -> {
+		void onWriteBufferFlushTimer() {
+			this.asyncLock.acquireThenCompose(this.writeTimeout, () -> {
 				return this.flushCoreAsync(FlushReason.TIMER, this.writeTimeout);
-			});
+			}).join();
```

Comment:
>).join(); [](start = 4, length = 9)

I need to go right now but this scares me. Can join be avoided?

Reply:
Can you describe the race which requires joining or not being async? The only thing I can think of would be if the async method hasn't completed yet and the timer fired again but I see how that can happen.

---
In reply to: [276024831](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276024831) [](ancestors = 276024831)

Reply:
It could be avoided, it's just that nobody checks on it for any exceptions. I was thinking that this join() shouldn't matter as much since this is on its own separate thread.

---
In reply to: [276024831](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276024831) [](ancestors = 276024831)

Reply:
With nothing joining or waiting here, the lock area exits after this.flushCoreAsync is called, not after it finishes executing. So isn't it possible that closeAsync() is called while flushCoreAsync() is still executing?

Maybe joining on the this.flushCoreAsync() inside is more appropriate?

---
In reply to: [276314775](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276314775) [](ancestors = 276314775,276024831)

Reply:
Removing the join(). It turns out to be some other issue which will be fixed later.

---
In reply to: [276314775](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276314775) [](ancestors = 276314775,276024831)

Reply:
It's pretty common for Timers to be implemented where thousands of timers get fired using the same thread (since creating then destroying a thread is historically relatively expensive).

If there are hundreds of other timers which need to be signaled while this async operation is in progress I don't want those timers to be delayed.

What if one of those timers was attempting to cancel this async operation because it was taking too long? That would be a form of a deadlock.

---
In reply to: [276352552](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276352552) [](ancestors = 276352552,276024831)

Reply:
I was anticipating seeing another iteration on this. Is that coming?

---
In reply to: [276738473](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276738473) [](ancestors = 276738473,276352552,276024831)